shell/shared/grain.js

```diff
@@ -386,6 +389,57 @@ Meteor.methods({
       return outerResult;
     }
   },
+  requestAccess: function (origin, grainId, identityId) {
+    check(origin, String);
+    check(grainId, String);
+    check(identityId, String);
+    if (!this.isSimulation) {
+      const grain = Grains.findOne(grainId);
+      if (!grain) {
+        throw new Meteor.Error(404, "No such grain");
+      }
+      const grainOwner = globalDb.getUser(grain.userId);
+      const email = _.findWhere(SandstormDb.getUserEmails(grainOwner), {primary: true});
+      if (!email) {
+        throw new Meteor.Error("no email", "Grain owner has no email address.");
+      }
+      const emailAddress = email.email;
+
+      const identity = globalDb.getIdentity(identityId);
+      globalDb.addContact(grainOwner._id, identityId);
+
+      const fromEmail = "Sandstorm server <no-reply@" + HOSTNAME + ">";
+      const senderEmails = SandstormDb.getVerifiedEmails(identity);
+      const replyTo = senderEmails.length > 0 ? senderEmails[0] : fromEmail;
+
+      // TODO(soon): In the HTML version, we should display an identity card.
+      let identityNote = "";
+      if (identity.profile.service === "google") {
+        identityNote = " (" + identity.privateIntrinsicName + " on Google)";
+      } else if (identity.profile.service === "github") {
+        identityNote = " (" + identity.profile.intrinsicName + " on GitHub)";
+      } else if (contact.profile.service === "email") {
+        identityNote = " (" + identity.profile.intrinsicName + ")";
+      }
+
+      const message = identity.profile.name + identityNote +
```

Comment:
I wonder if it's a phishing risk to include the requester's display name. We do include the verified email/username immediately thereafter which means we provide enough information to recognize impostors but I bet a lot of people won't look closely enough. Arguably we should remove the display name and only show the email address / github username which we can verify.

OTOH, in order to request access the user already has to have obtained the grain link which is itself a barrier to malicious users sending a request at all. Maybe at this point the usability benefit outweighs the phishing concern?

Thoughts? @zarvox interested in your take as well.